### PR TITLE
🐛 If resource marked as created, don't try create

### PR DESCRIPTION
**Pull request by grdryn**
## Overview

JIRA: https://issues.redhat.com/browse/INTLY-5819

If a resource was created on a cloud provider, but later was
accidentally deleted (or appears non-existent for some other reason),
the operator shouldn't try to create it again: manual intervention may
be needed (respond to alert, then restore from backup or delete CR).

This flag has been added as an annotation rather than a field in the
status block so that it can be manipulated manually. There are a
couple of cases where a human may want to do this:

- Restore an accidentally deleted CR that has/had a backing resource
on a cloud provider
- Migrate CR to a new cluster
- Force operator to create a new resource if restoration from backup
is not desirable

## Verification (updated with AWS-specific info)

Deploy the operator from this branch & configure to be able to create resources on AWS.

- Create one of each of Redis/Postgres/BlobStorage on AWS and make sure they all get created as expected, and then the "resourceIdentifier" annotation is set
- Maybe try deleting the backing resource for each of those CRs, and make sure that the operator doesn't recreate it.

- Create a CR for each of those CRDs with the annotation preset, and notice that the operator won't try to create anything for it.

### CODE REVIEW

File: pkg/controller/postgres/postgres_controller.go [hunk 1]
```diff
@@ -194,6 +203,11 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{Requeue: true, RequeueAfter: p.GetReconcileTime(instance)}, nil
 		}
 
+		annotations.Add(instance, annotations.Created, annotations.Created)
+		if err := r.client.Update(ctx, instance); err != nil {
+			return reconcile.Result{}, err
+		}
+
 		// return the connection secret
 		if err := r.resourceProvider.ReconcileResultSecret(ctx, instance, ps.DeploymentDetails.Data()); err != nil {
```
Comment (grdryn) [inline pkg/controller/postgres/postgres_controller.go]: This "ReconcileResultSecret" will now not happen if the "created" annotation exists (or the pre-existing case where `skipCreate` is specified on the spec. I imagine this is fine (`ps` won't exist here unless we do a create (unless there's another function that would "get" than existing one?), but I thought I'd highlight it anyway.

**Comment by R-Lawton** [comment]
:eyes:

File: pkg/annotations/annotations.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package annotations
```
[the remark this answers is not in the document]
Comment (ciaranRoche) [reply pkg/annotations/annotations.go]: I would lean toward putting this in `pkg/resources/annotations.go`

File: pkg/annotations/annotations.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Created is a string that can be used as an annotation
+const Created string = "created"
```
[the remark this answers is not in the document]
Comment (ciaranRoche) [reply pkg/annotations/annotations.go]: I would be happy to leave it here, over duplicating it across all controllers

File: pkg/controller/postgres/postgres_controller.go [hunk 4]
```diff
@@ -176,6 +177,14 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{Requeue: true, RequeueAfter: p.GetReconcileTime(instance)}, nil
 		}
 
+		if annotations.Has(instance, annotations.Created, annotations.Created) {
+			r.logger.Info("'created' annotation found, skipping postgres reconcile")
```
[the remark this answers is not in the document]
Comment (ciaranRoche) [reply pkg/controller/postgres/postgres_controller.go]: I lean more towards as much logging as possible. Looking at the logging that has been done in [cluster service](https://github.com/integr8ly/cluster-service/pull/3/files#diff-5d6e6baf22f72630659992325db5e91dR44) I think we should adopt this approach, and look to refactoring the logging. 

So in short for now, I would leave in as much logging as possible, allowing us to come back to refactor. Just my two cents

File: pkg/controller/postgres/postgres_controller.go [hunk 5]
```diff
@@ -176,6 +177,14 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{Requeue: true, RequeueAfter: p.GetReconcileTime(instance)}, nil
 		}
 
+		if annotations.Has(instance, annotations.Created, annotations.Created) {
```
[the remark this answers is not in the document]
Comment (ciaranRoche) [reply pkg/controller/postgres/postgres_controller.go]: I would leave as is. `SkipCreate` should IMO take priority as we do not want CRO to take any action on a resource when this is enabled. Allowing for human intervention with the resources with out CRO changing it

File: pkg/controller/postgres/postgres_controller.go [hunk 6]
```diff
@@ -194,6 +203,8 @@ func (r *ReconcilePostgres) Reconcile(request reconcile.Request) (reconcile.Resu
 			return reconcile.Result{Requeue: true, RequeueAfter: p.GetReconcileTime(instance)}, nil
 		}
 
+		annotations.Add(instance, annotations.Created, annotations.Created)
```
[the remark this answers is not in the document]
Comment (ciaranRoche) [reply pkg/controller/postgres/postgres_controller.go]: Yes, I would like to see a mapping here of the resource ID returned from AWS to the CR.
Not sure if this is in the scope of your current ticket, and I would not like to block the PR by adding this change. I will leave it up to yourself, depending on how much work it would take. 

Another option would be creating a JIRA for this, as you have laid the foundation here

File: pkg/controller/postgres/postgres_controller.go (hunk 1 above)
Comment (ciaranRoche) [reply pkg/controller/postgres/postgres_controller.go]: i would imagine this is okay, our alerts will highlight if anything is wrong. And for the most part after a restore the secret should get updated.

File: pkg/controller/postgres/postgres_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-martin) [reply pkg/controller/postgres/postgres_controller.go]: > Logs consume pod memory

@grdryn I'm interested in the rationale behind this comment.
I'm not holding up the PR or suggesting any changes, but I would like to know if there's more to this than my simplistic understanding.
I would think writing logs takes up some nominal amount of cpu & memory while writing, with most resource usage being filesystem for writing them somewhere. If a container is spewing out logs continuously at a rapid pace, I would understand an increase in memory usage, but nothing in the realm of anything significant so as to get the pod killed. If anything, cpu & storage would be the things that ramp up in usage.

Am I missing something?

[the remark this answers is not in the document]
Comment (grdryn) [reply pkg/controller/postgres/postgres_controller.go]: @david-martin I could be wrong about it, but we hit memory limits after a couple of days in the UPS operator because of the spammy logs (caused by an amq-online issue). I guess it could be due to something else, but our determination at the time was that it was holding them in memory and that's what was using all of the memory. It wasn't a big issue really...it just meant that the pod was recycled more frequently than we were expecting.

@ciaranRoche @david-martin the real reason I bring it up here is that this is the "normal" case, where the initial creation is a "special" case. Even for the happy path, these CRO controllers requeue a new reconcile request every <duration> (`p.GetReconcileTime(instance)`). That means that there will be a bunch of these "nothing to see here" messages (at least one per CR per <duration>). That's probably fine, and I'm probably just rambling. :slightly_smiling_face:

File: pkg/annotations/annotations.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (grdryn) [reply pkg/annotations/annotations.go]: I definitely can, but I think the functions might need to be renamed in that case. If I moved the file to that location, the code would be part of the `resources` package, and `resources.Add()` and `resources.Has()` doesn't really indicate what they do any longer. I guess `resources.AddAnnotation()` and `resources.HasAnnotation()` might be acceptable.

How would you describe the `resources` package? There seems to be a lot of stuff in there.

File: pkg/controller/postgres/postgres_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (grdryn) [reply pkg/controller/postgres/postgres_controller.go]: I'll leave that for now: currently all of this logic is at the top level of the controllers, so that we wouldn't have to think about it when adding future providers. I'm not sure what provider-specific info is available at this level. E.g. I don't know how to get the AWS resource ID here, or whatever the equivalent identifier for an in-cluster postgresql would be is.

**Comment by grdryn** [comment]
Currently working on getting the e2e tests to pass here.

File: pkg/providers/aws/provider_postgres.go [hunk 7]
```diff
@@ -189,11 +190,19 @@ func (p *PostgresProvider) createRDSInstance(ctx context.Context, cr *v1alpha1.P
...
-	if foundInstance == nil {
+	if foundInstance == nil && !annotations.Has(cr, resourceIdentifierAnnotation) {
 		logrus.Info("creating rds instance")
-		if _, err = rdsSvc.CreateDBInstance(rdsCfg); err != nil {
+		out, err := rdsSvc.CreateDBInstance(rdsCfg)
+		if err != nil {
 			return nil, croType.StatusMessage(fmt.Sprintf("error creating rds instance %s", err)), err
 		}
+
+		if out != nil && out.DBInstance != nil && out.DBInstance.DBInstanceIdentifier != nil {
```
Comment (grdryn) [inline pkg/providers/aws/provider_postgres.go]: It might be better to error in the cases where any of these are nil? If so, I'd need to figure out how to make the unit tests work in that case, since they'd be nil by default.

Maybe I can get them from the `rdsCfg` that's passed as input instead...I'll try that :thinking:

**Comment by grdryn** [comment]
I've updated this to just be for the AWS cloud provider now. I've also update the verification steps in the PR description above.

**Comment by ciaranRoche** [comment]
:eyes:

File: pkg/providers/aws/provider_postgres.go [hunk 8]
```diff
@@ -189,11 +190,16 @@ func (p *PostgresProvider) createRDSInstance(ctx context.Context, cr *v1alpha1.P
 	}
 
 	// create rds instance if it doesn't exist
-	if foundInstance == nil {
+	if foundInstance == nil && !annotations.Has(cr, resourceIdentifierAnnotation) {
```
[the remark this answers is not in the document]
Comment (grdryn) [reply pkg/providers/aws/provider_postgres.go]: Yeah, this could be the nil pointer issue I just discovered when testing. Good catch :+1:

File: pkg/providers/aws/provider_blobstorage.go [hunk 9]
```diff
@@ -404,6 +405,11 @@ func (p *BlobStorageProvider) reconcileBucketCreate(ctx context.Context, s3svc s
 		}
 	}
 	if foundBucket != nil {
+		if annotations.Has(bs, resourceIdentifierAnnotation) {
+			errMsg := fmt.Sprintf("BlobStorage CR %s in %s namespace has %s annotation with value %s, but no corresponding S3 Bucket was found",
+				bs.Name, bs.Namespace, resourceIdentifierAnnotation, bs.ObjectMeta.Annotations[resourceIdentifierAnnotation])
+			return croType.StatusMessage(errMsg), fmt.Errorf(errMsg)
```
Comment (grdryn) [inline pkg/providers/aws/provider_blobstorage.go]: Odd thing: this is now being hit on my cluster for some reason, for a bucket that should exist already.

Also, in the status block of the CR, the message seems to be shown 3 times:

```
status:
    message: 'BlobStorage CR example-blobstorage in cloud-resource-operator namespace
      has resourceIdentifier annotation with value gryandvd2dcloudresourceoperatorexam-2dwz,
      but no corresponding S3 Bucket was found: BlobStorage CR example-blobstorage
      in cloud-resource-operator namespace has resourceIdentifier annotation with
      value gryandvd2dcloudresourceoperatorexam-2dwz, but no corresponding S3 Bucket
      was found: BlobStorage CR example-blobstorage in cloud-resource-operator namespace
      has resourceIdentifier annotation with value gryandvd2dcloudresourceoperatorexam-2dwz,
      but no corresponding S3 Bucket was found'
```

Comment (grdryn) [reply pkg/providers/aws/provider_blobstorage.go]: > Odd thing: this is now being hit on my cluster for some reason, for a bucket that should exist already.

Well I fixed this issue at least. The problem was that I can't read: I had it in the `if foundBucket != nil` block, when I thought that said `if foundBucket == nil`.
